trie/stacktrie.go

```diff
@@ -0,0 +1,569 @@
+		// is an integer < 128
+		valHeaderLen = 0
+	}
+	if len(val) > 56 {
+		valHeaderLen = 2
+	}
+
+	// Add the global header, with optional length, and specify at
+	// which byte the header is starting.
+	payloadSize := int(keyByteSize) + (len(header) - headerPos - 1) +
+		valHeaderLen + len(val) /* value + rlp header */
+	var start int
+	if payloadSize >= 56 {
+		header[headerPos] = byte(payloadSize)
+		headerPos--
+		header[headerPos] = 0xf8
+		start = headerPos
+	} else {
+		header[headerPos] = 0xc0 + byte(payloadSize)
+		start = headerPos
+	}
+
+	// Write the header into the sponge
+	writer.Write(header[start:])
+
+	// Write the key into the sponge
+	var m byte
+	for i, nibble := range key {
+		// Skip the first byte if the key has an odd-length, since
+		// it has already been written with the header.
+		if i >= keyOffset {
+			if (i-keyOffset)%2 == 0 {
+				m = nibble
+			} else {
+				writer.Write([]byte{m*16 + nibble})
+			}
+		}
+	}
+
+	// Write the RLP prefix to the value if needed
+	if len(val) > 56 {
+		writer.Write([]byte{0xb8, byte(len(val))})
+	} else if len(val) != 1 || val[0] >= 128 {
+		writer.Write([]byte{0x80 + byte(len(val))})
+	}
+	writer.Write(val)
+
+	// DEBUG don't remove yet
+	//if leaf {
+	//fmt.Println("leaf rlp ", writer)
+	//} else {
+	//fmt.Println("ext rlp ", writer)
+	//}
+	//io.Copy(w, &writer)
+}
+
+func (st *StackTrie) hash() []byte {
+	/* Shortcut if node is already hashed */
+	if st.nodeType == hashedNode {
+		return st.val
+	}
+
+	var preimage bytes.Buffer
+	d := sha3.NewLegacyKeccak256()
+	switch st.nodeType {
+	case branchNode:
+		payload := [544]byte{}
+		pos := 3 // maximum header length given what we know
+		for i, v := range st.children {
+			if v != nil {
+				// Write a 32 byte list to the sponge
+				childhash := v.hash()
+				if len(childhash) == 32 {
+					payload[pos] = 128 + byte(len(childhash))
+					pos++
+				}
+				copy(payload[pos:pos+len(childhash)], childhash)
+				pos += len(childhash)
+				st.children[i] = nil // Reclaim mem from subtree
+			} else {
+				// Write an empty list to the sponge
+				payload[pos] = 0x80
+				pos++
+			}
+		}
+		// Add an empty 17th value
+		payload[pos] = 0x80
+		pos++
+
+		// Compute the header, length size is either 0, 1 or 2 bytes since
+		// there are at least 17 empty list headers, and at most 16 hashes
+		// plus an empty header for the value.
+		var start int
+		if pos-3 < 56 {
+			payload[2] = 0xc0 + byte(pos-3)
+			start = 2
+		} else if pos-3 < 256 {
+			payload[2] = byte(pos - 3)
+			payload[1] = 0xf8
+			start = 1
+		} else {
+			payload[2] = byte(pos - 3)
+			payload[1] = byte((pos - 3) >> 8)
+			payload[0] = 0xf9
+			start = 0
+		}
+
+		// Do not hash if the payload length is less than 32 bytes
+		if pos-start < 32 {
+			// rlp len < 32, will be embedded
+			// into its parent.
+			return payload[start:pos]
+		}
+		preimage.Write(payload[start:pos])
+	case extNode:
+		ch := st.children[0].hash()
+		if (len(st.key)/2)+1+len(ch) < 29 {
```

Comment:
size = "key" + "HP" + "keyHeader" + "val" + ("valueHeader" is unnecessary) + "listHeader"(at least 2)

len(st.key)/2)+1+len(ch) = "key" + "HP" + "val" 

If len(st.key)/2)+1+len(ch) == 29; e.g. "key" + "HP" = 1, "keyHeader" = 0, "val" = 28, "listHeader" = 2, then total size is 31, it's unnecessary to hash.